src/lib/optimizer/strategy/abstract_rule.cpp

```diff
@@ -14,4 +14,6 @@ void AbstractRule::_apply_to_inputs(std::shared_ptr<AbstractLQPNode> node) const
   if (node->right_input()) apply_to(node->right_input());
 }
 
+bool AbstractRule::prevents_caching() const { return false; }
```

Comment:
One more thing that came up when I wrote the previous response:

We should (1) document the relevance of this for the optimization, (2) explain how one would decide whether a new rule prevents caching or not, and (3) consider making this pure virtual, forcing every rule to implement it (and thus the developer to at least briefly think about it).